mlflow/types/utils.py

```diff
@@ -78,6 +87,185 @@ def _get_str_or_byte_type(data):
         return DataType.binary
 
 
+def _infer_datatype(data) -> Optional[DataType]:
+    if DataType.is_boolean(data):
+        return DataType.boolean
+    if DataType.is_integer(data):
+        return DataType.integer
+    if DataType.is_long(data):
+        return DataType.long
+    if DataType.is_float(data):
+        return DataType.float
+    if DataType.is_double(data):
+        return DataType.double
+    if DataType.is_string(data):
+        return DataType.string
+    if DataType.is_binary(data):
+        return DataType.binary
+    if DataType.is_datetime(data):
+        return DataType.datetime
+
+
+def _merge_object_properties(obj1: Object, obj2: Object) -> Object:
+    """
+    Check if two objects are compatible and return the updated object.
+    When we infer the signature from a list of objects, it is possible
+    that one object has more properties than the other. In this case,
+    we should mark those optional properties as required=False.
+    For properties with the same name, we should check the compatibility
+    of two properties and update.
+    An example of two compatible objects:
+
+        .. code-block:: python
+
+            obj1 = Object(
+                properties=[
+                    Property(name="a", dtype=DataType.string),
+                    Property(name="b", dtype=DataType.double),
+                ]
+            )
+            obj2 = Object(
+                properties=[
+                    Property(name="a", dtype=DataType.string),
+                    Property(name="c", dtype=DataType.boolean),
+                ]
+            )
+            updated_obj = _merge_object_properties(obj1, obj2)
+            assert updated_obj == Object(
+                properties=[
+                    Property(name="a", dtype=DataType.string),
+                    Property(name="b", dtype=DataType.double, required=False),
+                    Property(name="c", dtype=DataType.boolean, required=False),
+                ]
+            )
+
+    """
+    if obj1 == obj2:
+        return obj1
+    properties1 = {prop.name: prop for prop in obj1.properties}
+    properties2 = {prop.name: prop for prop in obj2.properties}
+    updated_properties = []
+    # For each property in the first element, if it doesn't appear
+    # later, we update required=False
+    for k in properties1.keys() - properties2.keys():
+        properties1[k].required = False
+        updated_properties.append(properties1[k])
+    # For common keys, property type should be the same
+    for k in properties1.keys() & properties2.keys():
+        updated_properties.append(_update_property_type(properties1[k], properties2[k]))
+    # For each property appears in the second elements, if it doesn't
+    # exist, we update and set required=False
```

Comment:
Can we use [symmetric_difference()](https://stackoverflow.com/questions/50871370/python-sets-difference-vs-symmetric-difference)?

Reply:
Ah wait, does this looks more complicated?
```
C = []
for k in A.keys().symmetric_difference(B.keys()):
    if k in A.keys():
        A[k].required = False
        C.append(A[k])
    else:
        B[k].required = False
        C.append(B[k])
```

Reply:
you're right. `symmetric_difference` seems more complicated.